Secure device session transfers with Noise (#134)
vacp2p/research: noise-research/secure-transfer.md

# Secure Transfers with Noise

In this document we describe a compound protocol to enable two devices to mutually authenticate and securely exchange (arbitrary) information. 
It consists of two main subprotocols or *phases*:

- [Device Pairing](#Device-Pairing): the devices exchange and authenticate their long term device ID static keys;
- [Secure Transfer](#Secure-Transfer): the devices securely exchange in encrypted form information using key material obtained during a successful pairing phase.

## Device Pairing

In the pairing phase, a device `B` requests to be paired to a device `A`. Once the two devices are paired, the devices will be mutually authenticated and will share a Noise session within which they can securely exchange information.

> The requests is made by exposing a QR code that, by default, has to be scanned by device `A`. 

Reviewer: ```suggestion
The request is made by exposing a QR code that, by default, has to be scanned by device `A`. 
```

If device `A` doesn't have a camera while device `B` does, [it is possible](#Rationale) to execute a slightly different pairing (with same security guarantees), where `A` is exposing a QR code instead.

### Employed Cryptographic Primitives

- `H`: the underlying hash function, i.e. SHA-256;
- `HKDF`: a key derivation function (based on SHA-256);
- `Curve25519`: the underlying elliptic curve for Diffie-Hellman (DH) operations.

### The `WakuPairing` Noise Handshake
The devices execute a custom handshake derived from `X1X1`, where they mutually exchange and authenticate their device static keys, i.e.

```
WakuPairing:
0.   <- eB              {H(sB||r), contentTopic}
     ...
1.   -> eA, eAeB        {H(sA||s)}   [auth_code]
2.   <- sB, eAsB        {r}
3.   -> sA, sAeB, sAsB  {s}

{}: payload,    []: user interaction
```

Beside the ephemeral key, all the information embedded in the QR code should be passed to the prologue of the Noise handshake (e.g. `H(sB||r)`, `contentTopic`, etc.).

### Protocol Flow
1. The device `B` exposes through a QR code a Base64 serialization of:
    - An ephemeral public key `eB`;

>     - A `contentTopic` where the information exchange will take place. `contentTopic` follows [23/WAKU2-TOPICS](https://rfc.vac.dev/spec/23/#content-topics) specifications and is then of the form `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/{random-message-id}/proto` for a randomly generated `random-message-id`. `contentTopic` can be then serialized in compressed form as `{application-name}:{application-version}:{shard-id}:{random-message-id}`.

Reviewer: What is the shard id?

Reviewer 2: > Why do we need a message id in the content topic? I'd avoid it if possible.

@s1fr0 could explain better, but afaik this is generated (or can be generated) for each message and allows for (only) the two parties to identify the next message to decrypt. Reasoning here: https://github.com/status-im/nwaku/issues/1081. However, this reminds me that we'd probably need to call it something other than `message-id`, as this concept already exists in libp2p for deduplication and is generated for each message on the GossipSub layer.

    - A commitment `H(sB||r)` for its static key `sB` where `r` is a random fixed-lenght value.

2. The device `A`:
    - scans the QR code;
    - obtains `eB`, `contentTopic`, `Hash(sB|r)`;
    - initializes the Noise handshake by passing `contentTopic` and `Hash(sB||r)` to the handshake prologue;
    - executes the pre-handshake message, i.e. processes the key `eB`;
    - executes the first handshake message over `contentTopic`, i.e. 
        - processes and sends a Waku message containing an ephemeral key `eA`; 
        - performs `DH(eA,eB)` (which computes a symmetric encryption key);
        - attach as payload to the handshake message a commitment `H(sA|s)` for `A`'s static key `sA`, where `s` is a random fixed-lenght value;
    - an 8-digits authorization code `auth_code` obtained as `HKDF(h) mod 10^8` is displayed on the device, where `h`is the handshake value obtained once the first handshake message is processed.

3. The device `B`:

>     - subscribes to any message sent to `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/*` and locally filters only messages sent to `contentTopic`. If any, continues.

Reviewer: > Another thing is that `filter` protocol does not yet support subscribing to partial content topic matches like `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/*`, but that shouldn't be a blocker for secure transfer.

Good point, AFAIK, store does not allow retrieval either of partial/globbing topics.

Same applies to any js/nim/go API to process inbound messages: a content topic `string` is expected, not a glob or regex.

Why do we need a message id in the content topic? I'd avoid it if possible.

>     - subscribes to any message sent to `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/*` and locally filters only messages sent to `contentTopic`. If any, continues.

Reviewer: > If this is to become a spec (which I think it should :) ), wondering if it's not better to avoid prescribing things like "subscribe" and "filter".

I agree.I would suggest using _listen to `contentTopic`_ for example.

It consists of two main subprotocols or *phases*:

- [Device Pairing](#Device-Pairing): the devices exchange and authenticate their long term device ID static keys;
- [Secure Transfer](#Secure-Transfer): the devices securely exchange in encrypted form information using key material obtained during a successful pairing phase.

## Device Pairing

In the pairing phase, a device `B` requests to be paired to a device `A`. Once the two devices are paired, the devices will be mutually authenticated and will share a Noise session within which they can securely exchange information.

The request is made by exposing a QR code that, by default, has to be scanned by device `A`. 
If device `A` doesn't have a camera while device `B` does, [it is possible](#Rationale) to execute a slightly different pairing (with same security guarantees), where `A` is exposing a QR code instead.

### Employed Cryptographic Primitives

- `H`: the underlying hash function, i.e. SHA-256;
- `HKDF`: a key derivation function (based on SHA-256);
- `Curve25519`: the underlying elliptic curve for Diffie-Hellman (DH) operations.

### The `WakuPairing` Noise Handshake

> The devices execute a custom handshake derived from `X1X1`, where they mutually exchange and authenticate their device static keys by exchanging messages over the content topic

Reviewer 3: If I understood correctly, the paring happens by scanning QR codes, right? then why we need a contentTopic? it is not over a p2p network?

```
contentTopic = /{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/proto
```

The handshake, detailed in next section, can be summarized as:

```
WakuPairing:
0.   <- eB              {H(sB||r), contentTopicParams, messageNametag}
     ...
1.   -> eA, eAeB        {H(sA||s)}   [authcode]
2.   <- sB, eAsB        {r}
3.   -> sA, sAeB, sAsB  {s}

{}: payload,    []: user interaction
```

### Protocol Flow

> 1. The device `B` exposes through a QR code a Base64 serialization of:

Reviewer 3: How do these numbered steps map to the summarized figure above? The numbers do not match 7 steps vs 3
Update: I figured out the relation, but it is very hard at the first glance. More explanation can clarify the relation between the two.

    - An ephemeral public key `eB`;
    - The content topic parameters `contentTopicParams = {application-name}, {application-version}, {shard-id}`.
    -  A randomly generated 8-bytes long `messageNametag`. 
    - A commitment `H(sB||r)` for its static key `sB` where `r` is a random fixed-lenght value.

2. The device `A`:

    - scans the QR code;
    - obtains `eB`, `contentTopicParams`, `messageNametag`, `Hash(sB|r)`;
    - checks if `{application-name}` and `{application-version}` from `contentTopicParams` match the local application name and version: if not, aborts the pairing. 
    - initializes the Noise handshake by passing `contentTopicParams`, `messageNametag` and `Hash(sB||r)` to the handshake prologue;
    - executes the pre-handshake message, i.e. processes the key `eB`;
    - executes the first handshake message over `contentTopic`, i.e. 
        - processes and sends a Waku message containing an ephemeral key `eA`; 
        - performs `DH(eA,eB)` (which computes a symmetric encryption key);

>         - attach as payload to the handshake message a commitment `H(sA|s)` for `A`'s static key `sA`, where `s` is a random fixed-lenght value;

Reviewer 3: ```suggestion
        - attaches as payload to the handshake message a commitment `H(sA|s)` for `A`'s static key `sA`, where `s` is a random fixed-length value;
```

>         - attach as payload to the handshake message a commitment `H(sA|s)` for `A`'s static key `sA`, where `s` is a random fixed-lenght value;

Reviewer 3: What is the purpose of `r` and `s` in `H(sA|s)` and `H(sB|r)`, I guess they are acting as salt, but if the hashes are already encrypted, then why another level of randomness?

    - an 8-digits authorization code `authcode` obtained as `HKDF(h) mod 10^8` is displayed on the device, where `h`is the handshake value obtained once the first handshake message is processed.

3. The device `B`:

    - listens to messages sent to `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/proto` and locally filters only those with [Waku payload](https://rfc.vac.dev/spec/35/#abnf) starting with `messageNametag`. If any, continues.
    - initializes the Noise handshake by passing `contentTopicParams`, `messageNametag` and `Hash(sB||r)` to the handshake prologue;
    - executes the pre-handshake message, i.e. processes its static key `eB`;
    - executes the first handshake message, i.e.
        - obtains from the received message a public key `eA`. If `eA` is not a valid public key, the protocol is aborted.
        - performs `DH(eA,eB)` (which computes a symmetric encryption key);
        - decrypts the commitment `H(sA||s)` for `A`'s static key `sA`.

>     - an 8-digits authorization code `authcode` obtained as `HKDF(h) mod 10^8` is displayed on the device, where `h`is the handshake value obtained once the first handshake message is processed.

Reviewer 3: Decimal digits? 
> an 8-digits authorization code

4. Device `A` and `B` wait the user to confirm with an interaction (button press) that the authorization code displayed on both devices are the same. If not, the protocol is aborted.
    
5. The device `B`:

    - executes the second handshake message, i.e.
        - processes and sends his (encrypted) device static key `sB` over `contentTopic`;
        - performs `DH(eA,sB)` (which updates the symmetric encryption key);
        - attaches as payload the (encrypted) commitment randomness `r` used to compute `H(sB||r)`.

6. The device `A`:

>     - listens to messages sent to `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/proto` and locally filters only those with Waku payload starting with `messageNametag`. If any, continues.

Reviewer 3: Is there any specific reason you did not say ` listens to messages sent to  contentTopic` like in step 5 above?
```suggestion
    - listens to messages sent to `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/proto` and locally filters only those with Waku payload starting with `messageNametag`. If any, continues.
```

>     - obtains from decrypting the received message a public key `sB`. If `sB` is not a valid public key, the protocol is aborted.

Reviewer 3: ```suggestion
    - decrypts the received message and obtains the public key `sB`. If `sB` is not a valid public key, the protocol is aborted.
```

    - Computes `H(sB||r)` and checks if this value corresponds to the commitment obtained in step 2. If not, the protocol is aborted.
    - executes the third handshake message, i.e.
        - processes and sends his (encrypted) device static key `sA` over `contentTopic`;
        - performs `DH(sA,eB)` (which updates the symmetric encryption key);
        - performs `DH(sA,sB)` (which updates the symmetric encryption key);
        - attaches as payload the (encrypted) commitment randomness `s` used to compute `H(sA||s)`.
    - Calls Split() and obtains two cipher states to encrypt inbound and outbound messages.
    

7. The device `B`:

    - listens to messages sent to `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/proto` and locally filters only those with Waku payload starting with `messageNametag`. If any, continues.
    - obtains from decrypting the received message a public key `sA`. If `sA` is not a valid public key, the protocol is aborted.
    - performs `DH(sA,eB)` (which updates a symmetric encryption key);
    - performs `DH(sA,sB)` (which updates a symmetric encryption key);
    - decrypts the payload to obtain the randomness `s`. 

>     - Computes `H(sA||s)` and checks if this value corresponds to the commitment obtained in step 6. If not, the protocol is aborted.

Reviewer 3: ```suggestion
    - Computes `H(sA||s)` and checks if this value corresponds to the commitment obtained in the step 3. If not, the protocol is aborted.
```

The resulting handshake would then be:
```
WakuPairing2:
0.   -> eA              {H(sB||r), contentTopicParams, messageNametag}
     ...
1.   <- eB, eAeB        {H(sB||r)}   [authcode]
2.   <- sB, eAsB        {r}
3.   -> sA, sAeB, sAsB  {s}

{}: payload,    []: user interaction
```

## Security Analysis

### Assumptions

- The attacker is active, i.e. can interact with both devices `A` and `B` by sending messages over `contentTopic`.

- The attacker has access to the QR code, that is knows the ephemeral key `eB`, the commitment `H(sB||r)` and the `contentTopic` exposed by the device `B`.

- Devices `A` and `B` are considered trusted (otherwise the attacker will simply exfiltrate the relevant information from the attacked device).

- As common for Noise, we assume that ephemeral keys cannot be compromised, while static keys might be later compromised. However, we enforce in the pairing phase extra security mechanisms (i.e. use of commitments for static keys) that will prevent some attacks possible when ephemeral keys are weak or get compromised.
 
### Rationale

- The device `B` exposes a commitment to its static key `sB` because:

>     - if the private key of `eB` is weak or gets compromised, an attacker can impersonate `B` by sending in message `2` to device `A` his own static key and successfully complete the pairing phase. Note that being able to compromise `eB` is not contemplated by our security assumptions.

Reviewer 3: Can you please explain more on what are the practical implications of an attacker being able to successfully complete the pairing phase? What will happen to B then? does it mean the whole communication between A and B is interrupted by the attacker hence B cannot send messages to A? or does it mean the attacker sits between A and B communication and can listen to that?

>     - `B` cannot adaptively choose a static key based on the state of the Noise handshake at the end of message `1`, i.e. after the authentication code is confirmed. Note that device `B` is trusted in our security assumptions.

Reviewer 3: How B can benefit from such an adaptive selection of its static key? or do you mean an Attacker? And if B is already trusted, then I'd say there is no need to prevent such an attack (although the attack is yet not clear to me)

>     - Confirming the authentication code after processing message `1` will ensure that no Man-in-the-Middle (MitM) can send a static key different than `sB`.

Reviewer 3: How do they confirm the auth code is identical if they are far away?

- Device `A` opens a commitment to its static key at message `3.` because:
    - if device `B` doesn't abort the pairing, device `A` acknowledges that device `B` correctly received his static key `sA`, since `s` was encrypted under an encryption key derived from the static keys `sA` and `sB` and the genuine (due to the previous `authcode` verification) ephemeral keys `eA` and `eB`.

# Secure Transfer

Once the handshake is concluded, sensitive information can be exchanged using the encryption keys agreed during the pairing phase. If stronger security guarantees are required, some [additional tweaks](#Additional-Possible-Tweaks) are possible.

In the following subsections we report the details of applications which are currently under development, mainly in order to implement [35/WAKU2-SESSION]().

However, the pairing and transfer phases descriptions are designed to be application-agnostic, and should be flexible enough to mutually authenticate and allow secure communication of two devices over a distributed network of Waku2 nodes.

## N11M session management mechanism

> In this scenario, one of Alice's devices is already communicating with one of Bob's devices within an active Noise session, e.g. after a successful execution of a Noise handshake.

Reviewer 3: By `after a successful execution of a Noise handshake.` do you mean a pairing process or not?
